### DISCUSSION TITLE
Relation between schema definition and extensions

### DISCUSSION THREAD

**Discussion by fubhy** [remirror/remirror#272] [Q&A]
## Checklist

- [x] I have read the [**contributing**](https://github.com/remirror/remirror/blob/master/docs/contributing.md) document.

## Question

Hi there!

I've discovered this project while researching the different options for ProseMirror / React bridges. Now, I've finally had the time and spent a few days this past week to see if Remirror would be a viable alternative for my use case. 

NOTE: I've been building on top of a customized Editor based on Atlaskit so far.

First and foremost: I am very intrigued by the outstanding code quality of this project. Great job!

While researching the code base, however, I noticed that one of the conclusions I drew (design rules that I extrapolated from my research so far) would be violated if I was to build my editor on top of Remirror. 

Namely: The extension system is tightly coupling the schema definition to the behavior of the editor.

I believe that any kind of ProseMirror Schema definition should be of purely semantic nature and not be tied to the interactive behavior of the editor at all.

I was wondering if you could share your thoughts around the relation between the ProseMirror Schema and the Remirror Extension API and why you decided to couple the two together.

Maybe it helps if I outline my reasoning of why I think coupling the two together is potentially problematic:

1. Ability to build tools simply and purely based on the schema
For instance, in my case, a command line tool for parsing and transforming legacy content into adf conforming to the schema. A simple, deterministic and stable schema from a plain .ts export seems to be desirable for such a case. Running the Extension Manager and loading the entire behavioral code seems counter-intuitive.

2. Deprecation of parts of the editor functionality without breaking legacy content
E.g. removing an extension but keeping the schema so that old content using the affected schema elements can still be rendered in an uneditable fashion.

3. Hidden implicit dependencies between extensions: For instance, the list item and bullet list  / ordered list extensions are tightly coupled on the schema level although they come as separate extensions. Same goes for tables and all other similar cases. On the schema level it makes no sense to separate those in the first place.

Instead of directly issuing the schema definitions in the extensions, it might work to instead pass the schema types as arguments to the extension constructor.

I am trying to understand the rational behind coupling them together. Would you mind sharing your thoughts around this?

Thank you, and thanks for this great project!

**Comment by ifiokjr** [comment]
You've raised some good points in this post. I think the main motivation for coupling the schema with the editor was to reduce the need for a deep prosemirror understanding when getting started. 

However with the `next` version of this API the issues raised can be addressed. I think there's space for providing an _optional_ way of predeclaring the schema that will be used. This can be passed as a prop to the editor component and used to validate whether the current setup meets the expected schema. A prop like `onSchemaFailure` could then be used to determines what happens when an unrecognised schema is encountered.

I'll explore this in the `next` branch and I'll let you know once it's open for contributions.

**Comment by fubhy** [comment]
Oh, right... It could actually be sufficient to simply separately export the default schema from each extension and then just pass it back to it when initializing it. 

> I'll explore this in the next branch and I'll let you know once it's open for contributions.

Great. I am looking forward to that. I am happy to engage more and contribute back if desirable.
